capabilities/src/execute_task_solution_capability.cpp

```diff
@@ -0,0 +1,189 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2016, Kentaro Wada.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Michael 'v4hn' Goerner */
+
+#include "execute_task_solution_capability.h"
+
+#include <moveit/plan_execution/plan_execution.h>
+#include <moveit/trajectory_processing/trajectory_tools.h>
+#include <moveit/kinematic_constraints/utils.h>
+#include <moveit/move_group/capability_names.h>
+#include <moveit/robot_state/conversions.h>
+
+namespace {
+
+// TODO: move to moveit::core::RobotModel
+const moveit::core::JointModelGroup* findJointModelGroup(const moveit::core::RobotModel& model, const std::vector<std::string>& joints) {
```

Comment:
This is an issue: The `Connect` stage creates robot trajectories that are merged from multiple `JointModelGroups`.
Hence, there doesn't exist a `JMG` for these trajectories!
However, I think `plan_execution` can handle them anyway: In https://github.com/ubi-agni/moveit_task_constructor/blob/automatica/core/python/wrapper/src/core.cpp#L102 I have used an arbitrary (the first) JMG to execute my trajectories. AFAIK the `ExecutionManager?` tries to find matching groups for any trajectory - that's the code you called over-engineered once ;-)

Reply:
Ok, things are a bit more ugly than I thought.

It is possible to create trajectories without a specified `JointModelGroup`.
But trajectory waypoints in `robot_trajectory::RobotTrajectory` are `RobotState`s which have no notion of which joints move anymore.
This is what the specified JMG is used for.

[`plan_execution` converts the `robot_trajectory::RobotTrajectory` objects to `moveit_msgs::RobotTrajectory`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning/plan_execution/src/plan_execution.cpp#L382)
to passes them to the `TrajectoryExecution` and [`robot_trajectory::RobotTrajectory::getRobotTrajectoryMsg`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_core/robot_trajectory/src/robot_trajectory.cpp#L200) assumes that all active joints might move and thus adds all of them to the trajectory message.
This is clearly not what should happen here but we need to modify `robot_trajectory::RobotTrajectory` to account for that - possibly adding a list of active joints to the object.

Your "trick", using a random group, only works because the trajectories are forwarded to the `TrajectoryExecutionManager` as `moveit_msgs::RobotTrajectory` there, no matter which group is associated with the MGI. - Btw this also entails that these trajectories are not validated online.

For the plan execution the group actually matters as described above.

Until the `RobotTrajectory` class is changed, I propose to leave this as is and define auxiliary groups. In your case you can simply define meta-groups that contain the groups that might be merged in the srdf. I don't like the additional complexity of sets of groups here because it adds an unnecessary layer of abstraction.